libnavigation-core/src/main/java/com/mapbox/navigation/core/internal/trip/session/MapboxTripSession.kt

```diff
@@ -401,37 +399,27 @@ class MapboxTripSession(
     }
 
     private fun updateRawLocation(rawLocation: Location) {
+        unconditionalStatusPollingJob?.cancel()
         locationObservers.forEach { it.onRawLocationChanged(rawLocation) }
         ioJobController.scope.launch {
             val currentDate = Date()
-            lastStatusUpdateTimeMillis = SystemClock.elapsedRealtime()
             navigator.updateLocation(rawLocation, currentDate)
             updateDataFromNavigatorStatus(currentDate)
         }
 
-        if (this.rawLocation == null) {
-            ioJobController.scope.launch {
-                while (isActive) {
-                    delay(STATUS_POLLING_INTERVAL)
-                    launch {
-                        emitUnconditionallyGetStatus()
-                    }
+        unconditionalStatusPollingJob = ioJobController.scope.launch {
+            delay(UNCONDITIONAL_STATUS_POLLING_PATIENCE)
+            while (isActive) {
+                launch {
```

Comment:
do we need to use extra `launch` builder?

Reply:
I'm not sure, looks fine to remove. @korshaknn do you have any additional context?

Reply:
That's needed because if not the `delay`s (or any other operations) are executed in the same thread sequentially, affecting the overall frequency as it depends on the time that they take i.e. the polling won't be exactly respected because of this extra time that these operations add. We noticed that in the past and added these `launch`.

Reply:
`delay` is still not perfect in terms of achieving exactly 1s delay.

Reply:
@Guardiola31337 are there any tickets/PRs to read about it? It's still not crystal clear for me what the issue was

Reply:
Something that we can't control though, Kotlin implementation ¯\\\_(ツ)\_/¯ In any case is not critical for us if it takes a bit longer that a second but if for example the operation notifies listeners, that depends on the number of listeners registered and the time of that operation - could be a long operation that could cause that issues of missing loops (if it takes longer than the polling time).

Reply:
we launch a new coroutine each second. Maybe it's worth to remove the second `launch`?
 @Guardiola31337 I'm not sure that completely got your point. Could you please describe the whole flow `with = and without` inner `launch`?

Reply:
Yeah 👀 

I removed `delay(UNCONDITIONAL_STATUS_POLLING_PATIENCE)` for simplicity

Without `launch`

```kotlin
launch {
            while (isActive) {
                updateDataFromNavigatorStatus(Date())
                delay(UNCONDITIONAL_STATUS_POLLING_INTERVAL)
            }
}
```

- outer `launch` 👉 new `Job` that executes the operations sequentially
- Imagine that `updateDataFromNavigatorStatus(Date())` takes 3s
- `delay` will be executed at `t` 3s
- `delay` will stop 1s

That's an issue because we want `updateDataFromNavigatorStatus(Date())` to be executed every second and in this case will be executed every 4 seconds.

With `launch`

```kotlin
launch {
            while (isActive) {
                launch {
                    updateDataFromNavigatorStatus(Date())
                }
                delay(UNCONDITIONAL_STATUS_POLLING_INTERVAL)
            }
}
```

- outer `launch` 👉 new `Job` that executes the operations sequentially
- inner `launch` 👉 new `Job` that executes the operations sequentially but in a separate thread (not blocking)
- now `updateDataFromNavigatorStatus(Date())` can take 3s that doesn't affect the upcoming `delay` operation
- `delay` will be executed at `t` 0s
- `delay` will stop 1s

Doing this `updateDataFromNavigatorStatus(Date())` will be executed every second (if `delay` implementation would be perfect 😅 @evabishchevich) but it's not blocked by whatever is done in `updateDataFromNavigatorStatus(Date())` and it's irrespective of the time it takes.

Does that make sense?

Reply:
yeah, thx @Guardiola31337 . To be crystal clear - this case works as described if we have Dispatcher with a bunch of threads behind it. Otherwise
`inner launch 👉 new Job that executes the operations sequentially but in a separate thread (not blocking)` will block the thread and we will have `updateDataFromNavigatorStatus` executed every 3 secs (delay will be skipped due to implementation details if I'm not mistaken).

Reply:
@Guardiola31337 that makes sense! but we might get even more problems. 
`now updateDataFromNavigatorStatus(Date()) can take 3s that doesn't affect the upcoming delay operation`
yep, it means that we will launch the same method from different threads.
Do we have synchronization? Do we have shared mutable state?
```
private fun updateDataFromNavigatorStatus(date: Date) {
        mainJobController.scope.launch {
            val status = getNavigatorStatus(date)
            launch {
                updateEnhancedLocation(status.enhancedLocation, status.keyPoints)
                updateRouteProgress(status.routeProgress)
                isOffRoute = status.offRoute
            }
        }
    }
```
might we have any issues here when it is called from 2 or 3 threads?

Reply:
btw we have 2 `launch` blocks here too ^^

Reply:
> yeah, thx @Guardiola31337 . To be crystal clear - this case works as described if we have Dispatcher with a bunch of threads behind it. Otherwise
>
>> inner launch 👉 new Job that executes the operations sequentially but in a separate thread (not blocking)
>
> will block the thread and we will have `updateDataFromNavigatorStatus` executed every 3 secs (delay will be skipped due to implementation details if I'm not mistaken).

That's right @evabishchevich 👍 Actually our IO dispatcher has 2 threads 👀 https://github.com/mapbox/mapbox-navigation-android/blob/7f527d7a0455e1fa827dc169f722d207c35eeab5/libnavigation-util/src/main/java/com/mapbox/navigation/utils/internal/ThreadController.kt#L50-L57 That being said we haven't seen any contention issues with that number (https://github.com/mapbox/mapbox-navigation-android/issues/3211#issuecomment-648921165 maybe yes?) but yeah if that happens in the future that might be the root cause, agreed.

Reply:
Are we good here or any clear next actions?

Reply:
@korshaknn these are really great points! We might need to check if that could be an issue, could you take a quick stab at it? BTW let's cut a separate ticket to investigate this, this discussion / issue shouldn't block this PR.

> btw we have 2 `launch` blocks here too ^^

And yes, that's not the only place in which we follow that approach.